packages/components/src/presentation-components/properties/inputs/ItemsLoader.tsx

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @internal */
+export const VALUE_BATCH_SIZE = 100;
+
+/** @internal */
+export const FILTER_WARNING_OPTION = { label: "Too many values please use filter", value: "__filter__", disabled: true };
+
+interface LoadedItems<T> {
+  options: T[];
+  length: number;
+  hasMore: boolean;
+}
+
+/** @internal */
+export class ItemsLoader<T> {
+  private _loadedItems: T[] = [];
+  private _isLoading = false;
+  private _hasMore = true;
+  private _disposed = false;
+  private _offset = 0;
+
+  constructor(
+    private _beforeLoad: () => void,
+    private _onItemsLoaded: (newItems: T[]) => void,
+    private _loadItems: (offSet: number) => Promise<LoadedItems<T>>,
+    private _getOptionLabel: (option: T) => string,
+  ) {}
+
+  public dispose() {
+    this._disposed = true;
+  }
+
+  public async loadInitialItems(initialSelectedValues?: string[]) {
+    const matchingItems: T[] = [];
+
+    const needsItems = (): boolean => {
+      if (this._loadedItems.length > 0) {
+        filterInitialSelectedValues(this._loadedItems);
+      }
+
+      return this._loadedItems.length < VALUE_BATCH_SIZE && (initialSelectedValues?.length === 0 || initialSelectedValues?.length !== matchingItems.length);
+    };
+
+    const filterInitialSelectedValues = (items: T[]) => {
+      matchingItems.push(
+        ...items.filter((item) => {
+          return initialSelectedValues?.some((initialSelectedValue) => initialSelectedValue === this._getOptionLabel(item));
+        }),
+      );
+    };
+
+    const needsMoreItems = (items: T[]): boolean => {
+      filterInitialSelectedValues(items);
+      return !!initialSelectedValues && initialSelectedValues.length !== matchingItems.length;
+    };
+
+    await this.loadUniqueItems(needsItems, needsMoreItems);
+  }
+
+  public async loadItems(filterText?: string) {
+    const matchingItems: T[] = [];
+
+    if (!filterText) {
+      return;
+    }
+
+    const needItems = (): boolean => {
+      matchingItems.push(...this._loadedItems.filter((option) => this._getOptionLabel(option).toLowerCase().includes(filterText.toLowerCase())));
+      return matchingItems.length < VALUE_BATCH_SIZE;
+    };
+
+    const needsMoreItems = (options: T[]): boolean => {
+      matchingItems.push(...options.filter((option) => this._getOptionLabel(option).toLowerCase().includes(filterText.toLowerCase())));
+      return matchingItems.length < VALUE_BATCH_SIZE;
+    };
+
+    await this.loadUniqueItems(needItems, needsMoreItems);
+  }
+
+  private async loadUniqueItems(needsItems: () => boolean, needsMoreItems: (options: T[]) => boolean) {
```

Comment:
Looks like instead of `needsItems` you can call `needsMoreItems(this._loadedItems)`.

Reply:
Changed it to pass needsMoreItems instead and removed `needsItems` from `loadInitialItems`

Reply:
That's not exactly what I meant. I think `loadUniqueItems` function can take only `needsMoreItems` callback and call it instead of `needsItems()` like `needsMoreItems(this._loadedItems)`

Reply:
`loadInitialItems` also calls loadUniqueItems and merging `needsItems` with `needsMoreItems` would be pretty hard because their conditions don't match (needsItems check that if for some reason some items are loaded perhaps we don't need to load more, which would not work with `needsMoreItems`)

Reply:
Wouldn't something like this work:
```ts
const matchingItems: T[] = [];
const needsItemsLoaded = (items: T[]) => {
  if (initialSelectedValues && initialSelectedValues.length > 0) {
    matchingItems.push(
      ...items.filter((item) => {
        return initialSelectedValues.some((initialSelectedValue) => initialSelectedValue === this._getOptionLabel(item));
      }),
    );
    return matchingItems.length === initialSelectedValues.length;
  }
  matchingItems.push(...items);
  return matchingItems.length < VALUE_BATCH_SIZE;
}
```

Reply:
I don't think so, you would only want to have the exact items that have the same label in the matching items as you check for ===, but in the else we would just push all of them, we would also need to check against already loaded items if we want to have the initial load. I think it's possible to merge `needsItems` and `needsMoreItems`, but it would be inefitient as some if statements which do not matter for while cycle would be hit and the merged `needsMoreItems` would be unnecessary complicated.

Reply:
changed